aptos-move/mvhashmap/src/versioned_data.rs

```diff
@@ -259,21 +264,41 @@ impl<K: Hash + Clone + Debug + Eq, V: TransactionWrite> VersionedData<K, V> {
         let mut v = self.values.entry(key).or_default();
         let bytes_len = data.bytes_len();
         // For base value, incarnation is irrelevant, set to 0.
-        let prev_entry = v.versioned_map.insert(
-            ShiftedTxnIndex::zero(),
-            CachePadded::new(Entry::new_write_from(0, data, maybe_layout)),
-        );
 
-        assert!(prev_entry.map_or(true, |entry| -> bool {
-            if let EntryCell::Write(i, v, _) = &entry.cell {
-                // base value may have already been provided due to a concurrency race,
-                // but it has to be the same as being set.
-                // Assert the length of bytes for efficiency (instead of full equality)
-                *i == 0 && v.bytes_len() == bytes_len
-            } else {
-                true
-            }
-        }));
+        use btree_map::Entry::*;
+        match v.versioned_map.entry(ShiftedTxnIndex::zero()) {
+            Vacant(v) => {
+                v.insert(CachePadded::new(Entry::new_write_from(
+                    0,
+                    data,
+                    maybe_layout,
+                )));
+            },
+            Occupied(o) => {
+                assert!(
+                    if let EntryCell::Write(i, v, layout) = &o.get().cell {
+                        // base value may have already been provided due to a concurrency race.
+                        // If maybe_layout is None, they are required to be identical.
+                        // If maybe_layout is Some, there might have been an exchange
+                        // Assert the length of bytes for efficiency (instead of full equality)
+                        layout.is_some() == maybe_layout.is_some()
+                            && *i == 0
+                            && (layout.is_some() || v.bytes_len() == bytes_len)
```

Comment:
I didn't get it. We check that the length of the data matches only when layout is None. Why is that?

Reply:
added more comments, and explanation here - https://aptos-org.slack.com/archives/C057VPCD7QX/p1697520278313169